### ISSUE TITLE
Dart bindings leak access to capabilities

### ISSUE THREAD

**Issue by j9brown** [domokit/mojo#750]
/cc @jlnt, @krisgiesing 

The Dart bindings use raw integers to represent handles (wrapped up in a CoreHandle) and don't check whether a given isolate should have access to a handle before passing it to the Mojo APIs.

We can see this in MojoHandleNatives.wait() for example, which just takes an int from Dart land, passes it down to the natives, which cast it to MojoHandle and issue the system call.  As a consequence of this implementation, any isolate can use any MojoHandle available in the process so we lose capability isolation.

Here are some ideas for fixing this:

1. Represent MojoHandles as opaque Dart objects with the actual handle value stored in native code.  Never leak the integer handle to Dart and never allow Dart code to manufacture a MojoHandle from an integer.

2. Validate handles as they cross the boundary (expensive, unnecessary if we do #1 above).

To be honest, the more I look at the Dart bindings, the more I think we should scrap them and start over with a more deliberate focus on security, performance, and API usability.

**Comment by krisgiesing** [comment]
I think this brings up a philosophical issue of the purpose of isolates. The sandboxing aspect seems to arise from a model where isolates serve something like the purpose of a fork() call, but many use cases (including the one that spurred this discussion) are more like a replacement for threads. In addition, in a full Turquoise ecosystem it's not clear to me that the fork()-like functionality would be accomplished through a Dart isolate; it seems more likely that this would be accomplished through a call to a Mojo service, since that's a more language-neutral way to expose the capability.

So, before we do a bunch of work on this code, I think I'd like to know more about the scope of requirements and where they are coming from.

**Comment by krisgiesing** [comment]
FYI, a naive attempt to pass an integer across to another isolate, and then instantiate a mojo handle from that int, resulted in this exception:

E/flutter : [ERROR:dart_handle_watcher.cc(149)] (Dart Handle Watcher) Handle 30 is already bound!

So I think there is some safeguarding in place already.

**Comment by zanderso** [comment]
The Mojo message codec for the Dart bindings is currently implemented in Dart, so the raw handle integer values need to be available there. Making the raw handle values completely unavailable to Dart code would require rewriting the codec in C++ to build up the Dart objects for the decoded messages using Dart's C API. A less disruptive middle ground would be to try to isolate the raw handle values to mojo/core.dart and/or mojo/bindings.dart libraries using the '_' prefix on names.

@j9brown if you have specific issues with the Dart bindings, please file bugs here on github, otherwise no one is going to work on the problems. Just saying "Let's scrap it and start over" is not constructive.

**Comment by j9brown** [comment]
I agree that's not constructive.  Let's meet and develop a plan for the
evolution of these bindings.

**Comment by j9brown** [comment]
Incidentally, simply making classes and fields private is inadequate as a
security mechanism since they can still be accessed via reflection if
mirrors are enabled.  They may even be accessible directly if the private
symbol name prefix can be guessed.  I don't know what kind of binary
validation the Dart VM performs, particularly in unchecked mode.

In any case, this is a problem because some folks are considering using
isolates as a security boundary for running untrusted / semi-trusted code
within a shared process so we need to be on the lookout for ways that an
isolate could access capabilities it shouldn't have access to.
Synthesizing Mojo handles is one example of such a situation but there are
probably others lurking in the shadows.  (eg. How are file descriptors
wrapped up?  Are there any bare pointers that could be manipulated?  Are
there native methods that could be abused?  etc...)

(Aside: we'll also need a mechanism to safely transfer Mojo handles to
other isolates via messages or similar means...)

**Comment by zanderso** [comment]
I need to spend some more time thinking about this, but a few quick things:

1. The ability to try to forge handles seems implied by the Mojo philosophy of allowing a Mojo app to jump in at any abstraction layer it likes, i.e. at the system call layer (dart:mojo.internal), the mojo/core.dart layer, the mojo/bindings.dart layer, or the mojo/application.dart layer. Removing that capability would make Dart a second-class citizen from the perspective of this part of Mojo philosophy, at least as it's been explained to me. IOW, you could do stuff with Java or C++ that you couldn't do with Dart. But if that's what everybody wants, then we can certainly discuss moving in that direction.

2. Mirrors are disabled by both Mojo and Flutter, and there is no plan to re-enable them.

3. Attempts to forge a private symbol name are rejected by the Dart parser.

4. We may want to consider using https://api.dartlang.org/stable/1.15.0/dart-isolate/Capability-class.html to create capabilities for various things. At a coarse grain, a spawning isolate could pass the Mojo capability to isolates running trusted but not untrusted Dart scripts, and the calls in dart:mojo.internal could be modified to take and check the capability.

**Comment by j9brown** [comment]
I don't agree with #1.  Dart apps can still jump in at the syscall layer,
we just need to define that layer to prevent forging of handles.  Given
that Mojo handles are already transmitted separately from message bodies
(in another array), that should only require wrapping them up as Dart
objects safely as I proposed initially.

I don't think there's any reason for Dart code to be manipulating handles
as bare integers and nothing important will be lost by preventing this.

Moreover, even C++ apps won't be trusted to do things that could violate
the integrity of their sandbox, perhaps using SFI.  Dart is no different.
There must be strict validation for dangerous operations which may violate
the invariants of the sandbox (compromising the process or other isolates
running within it).

Incidentally, although the Dart parser forbids generating references to
private symbols, does the Dart runtime actually check accessibility rules
when it loads code?  A casual Google search suggested it might not.  So
through manipulation of Dart binaries one could essentially bypass access
control even without using mirrors.  (And who knows, someone might want to
reenable mirrors someday.)

In general it's unwise to rely on access control as a security boundary.
The VM and native code must enforce the sandbox invariants.  That's why I
suggested making Mojo handles opaque.

**Comment by j9brown** [comment]
Other things to worry about: Dart code shouldn't be able to manipulate raw
pointers to memory, perform arbitrary syscalls, or do other things which
might violate the sandbox invariants.  (Possible exception might be highly
trusted services but most code downloaded from the internet won't fit this
bill.)

**Comment by zanderso** [comment]
> So through manipulation of Dart binaries one could essentially bypass access control even without using mirrors.

I believe this is impossible. If you have a reproduction, please file a bug against the VM. /cc @iposva-google 

> Other things to worry about: Dart code shouldn't be able to manipulate raw
> pointers to memory, perform arbitrary syscalls, or do other things which
> might violate the sandbox invariants.

We've been pretty careful about this. Dart code never has access to raw pointers. If you find that it does, please file a bug against the VM. In the Mojo shell, the bits of `dart:io` that are supported are implemented via Mojo services, not by syscalls. If there is a doc describing the sandboxing invariants we will obviously need to go over that in greater detail before making big decisions or doing big refactorings.

**Comment by j9brown** [comment]
The essential sandboxing invariant is simple: Dart isolates shouldn't be able to snoop on or compromise the integrity of other code that happens to be running elsewhere in the same process.

We are not overly concerned with denial-of-service and covert side-channels but we do care very much about preventing direct information leakage, much like a web browser does.

**Comment by jamesr** [comment]
How are Dart programs supposed to pass handles between isolates?  If there is a way for the C++ side of the Dart bindings to observe handle transfers, it'd be easy for these bindings to verify that all handle values it observes in syscalls match handles that the calling isolate has access to.  All operations that transfer handle values around are explicit in the Mojo APIs so tracking those ops is not difficult.
